go/vt/tabletserver/grpctabletconn/conn.go

```diff
@@ -139,50 +138,51 @@ func (conn *gRPCQueryClient) ExecuteBatch(ctx context.Context, target *querypb.T
 	return sqltypes.Proto3ToResults(ebr.Results), nil
 }
 
-type streamExecuteAdapter struct {
-	stream queryservicepb.Query_StreamExecuteClient
-	fields []*querypb.Field
-}
+// StreamExecute starts a streaming query to VTTablet.
+func (conn *gRPCQueryClient) StreamExecute(ctx context.Context, target *querypb.Target, query string, bindVars map[string]interface{}, options *querypb.ExecuteOptions, sendReply func(*sqltypes.Result) error) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-func (a *streamExecuteAdapter) Recv() (*sqltypes.Result, error) {
-	ser, err := a.stream.Recv()
-	switch err {
-	case nil:
-		if a.fields == nil {
-			a.fields = ser.Result.Fields
+	stream, err := func() (queryservicepb.Query_StreamExecuteClient, error) {
+		conn.mu.RLock()
+		defer conn.mu.RUnlock()
+		if conn.cc == nil {
+			return nil, tabletconn.ConnClosed
 		}
-		return sqltypes.CustomProto3ToResult(a.fields, ser.Result), nil
-	case io.EOF:
-		return nil, err
-	default:
-		return nil, tabletconn.TabletErrorFromGRPC(err)
-	}
-}
-
-// StreamExecute starts a streaming query to VTTablet.
-func (conn *gRPCQueryClient) StreamExecute(ctx context.Context, target *querypb.Target, query string, bindVars map[string]interface{}, options *querypb.ExecuteOptions) (sqltypes.ResultStream, error) {
-	conn.mu.RLock()
-	defer conn.mu.RUnlock()
-	if conn.cc == nil {
-		return nil, tabletconn.ConnClosed
-	}
 
-	q, err := querytypes.BoundQueryToProto3(query, bindVars)
+		q, err := querytypes.BoundQueryToProto3(query, bindVars)
+		if err != nil {
+			return nil, err
+		}
+		req := &querypb.StreamExecuteRequest{
+			Target:            target,
+			EffectiveCallerId: callerid.EffectiveCallerIDFromContext(ctx),
+			ImmediateCallerId: callerid.ImmediateCallerIDFromContext(ctx),
+			Query:             q,
+			Options:           options,
+		}
+		stream, err := conn.c.StreamExecute(ctx, req)
+		if err != nil {
+			return nil, tabletconn.TabletErrorFromGRPC(err)
+		}
+		return stream, nil
+	}()
 	if err != nil {
-		return nil, err
-	}
-	req := &querypb.StreamExecuteRequest{
-		Target:            target,
-		EffectiveCallerId: callerid.EffectiveCallerIDFromContext(ctx),
-		ImmediateCallerId: callerid.ImmediateCallerIDFromContext(ctx),
-		Query:             q,
-		Options:           options,
+		return err
 	}
-	stream, err := conn.c.StreamExecute(ctx, req)
-	if err != nil {
-		return nil, tabletconn.TabletErrorFromGRPC(err)
+	var fields []*querypb.Field
+	for {
+		ser, err := stream.Recv()
+		if err != nil {
+			return tabletconn.TabletErrorFromGRPC(err)
+		}
+		if fields == nil {
+			fields = ser.Result.Fields
+		}
+		if err := sendReply(sqltypes.CustomProto3ToResult(fields, ser.Result)); err != nil {
+			return tabletconn.TabletErrorFromGRPC(err)
```

Comment:
Using `TabletErrorFromGRPC` is not correct here: The only error you can see here are provided by the application callback and not by gRPC.

Therefore, you don't need to convert any gRPC error.

The only reason you're calling this function is probably because you want to transform io.EOF to nil?

How about checking here explicitly for io.EOF? This way the behavior of the code is much clearer and it won't be a surprise that io.EOF will be "hidden".